Sources/CardPayments/CardClient.swift

```diff
@@ -154,6 +164,43 @@ public class CardClient: NSObject {
         return url.queryItems?.first { $0.name == param }?.value
     }
 
+    private func startVaultThreeDSecureChallenge(
+        url: URL,
+        setupTokenID: String
+    ) {
+
```

Comment:
I guess my question is more, do we only need different result delegates, 1 for vault, 1 for non-vault. It seems like most of the other delegates: 3dsWillLaunch, cancellation, failure shouldn't matter if the merchant is vaulting or not. And with that, does it make more sense to just add 3ds data to 1 result reguardless of vault data? We could potentially add `VaultResult` to the `CardResult` and return extra data there if vaulted.

It just seems to me like a lot of this logic is separated out when it may not need to be since under the hood both the vault and non-vault flow do the same thing: take in a string, launch a URL, and return a result. Curious to hear your thoughts though!

Reply:
I think we would need different delegates to return different results since they are different flows with different result values. id in CardResult is orderID and id in CardVaultResult is setupTokenID. And they need different state properties for different steps in the app for different endpoints, for example.

Reply:
I think we should also take into consideration the analytics we should add, I think it is better to have different analytics depending on the flow, I agree that using a single function for both flows would be correct, however I also think it could be confusing and prone to errors, we could do an alaisis, even divide the CardClient (thinking out loud) for the future, what do you think? Do you have any options to have a 3DS function for these two flows?

Reply:
Ideally we'd send vault vs non vault in some sort of metadata parameter on our analytics. This would allow us to better compare a module's overall performance as well as drilling down into specific performance of vault vs non-vaulted flows. With sending separate event names for each flow, we end up needing to add events together to get a more holistic picture of a modules conversion which also leaves us prone to missing events.

It seems like as a whole in other modules we've drawn a pretty strict line around vault vs non-vault, so this could be something we consider tackling in v2 but overall I think in cases where we have the same code duplicated in 2 places with only ~5 lines of code different there is likely a way we can DRY up the codebase. Both are doing the exact same thing: take in a string, launch a URL, and return a result. Any time we are in a situation where we end up needing to not forget code exists in 2 places for launching URLs we could potentially open ourselves up for errors if we need to make updates. Not sure we have to tackle it in this PR but worth exploring.

Reply:
Curious to hear @scannillo thoughts on this too when she is back!